testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/AutoCleanupExtension.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.testing.junit;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * A small utility extension that allows deferring executing cleanup code in {@code @Test} methods
+ * until the test has finished. All cleanup callbacks added during the test will always be executed
+ * after it finishes, no matter the outcome. Inspired by Spock's {@code cleanup:} block.
+ */
+public final class AutoCleanupExtension implements AfterEachCallback {
+  private final Queue<AutoCloseable> thingsToCleanUp = new LinkedList<>();
```

Comment:
```suggestion
  private final Queue<AutoCloseable> thingsToCleanUp = new ConcurrentLinkedQueue<>();
```

Reply:
Hm, I think JUnit creates a new test class instance for each test method and there's hardly any need to care about thread safety, but this can't hurt anyway.